File: packages/oos-dsls-ts/src/renderer/llm-chunk.ts
// renderer/llm-chunk.ts — DomainDef → LLM-friendly text chunk.
//
// Direct port of `renderChunk()` and its writer functions from
// `oosp/pluginsrv/store/schema_chunk.go`. The contract for downstream
// consumers (granite-embedding + pgvector + frontend LLM) is preserved:
// the same shape of chunk produces the same retrieval behaviour.
//
// Why this lives in the DSL package and not the backend: the renderer
// is a pure function (DomainDef → string) with no I/O. Keeping it
// next to the parser and mapper means the frontend LLM-debug tooling
// can call it directly to show "what does the system know about
// person?" without round-tripping to oosp.
//
// The chunk is composed in numbered sections, each emitted by a
// dedicated writer function. Order matters: the embedding model sees
// header → fields → filter examples → relations → metas → permissions
// → AI hints, in that order, every time. Stability across rebuilds
// is part of the contract.

import type {
	AiHintDef,
	DomainDef,
	DomainFieldDef,
	ExampleDef,
	FieldTypeDef,
	MetaDef,
	PermissionDef,
	RelationDef,
} from "../types";
import { domainAliases } from "./aliases";
import {
	findOperator,
	formatExampleValue,
	operatorsForType,
	renderFilterArg,
	type FilterOp,
} from "./operators";

/**
 * renderLLMChunk produces the structured, human-readable description
 * of a domain that gets embedded into pgvector for LLM retrieval.
 *
 * The output is deterministic: same DomainDef → same string, byte for
 * byte. That property lets the embedding pipeline detect "no change,
 * skip re-embed" by hashing the chunk text.
 */
export function renderLLMChunk(def: DomainDef): string {
	const parts: string[] = [];
	writeHeader(parts, def);
	writeAliases(parts, def);
	writeFields(parts, def);
	writeFilterExamples(parts, def);
	writeRelations(parts, def);
	writeMetas(parts, def);
	writeDropdownFieldMapping(parts, def);
	writePermissions(parts, def);
	writeAiHints(parts, def);
	return parts.join("");
}

// ─── Section: header ─────────────────────────────────────────────────

/** writeHeader emits the one-line identity of the domain. */
function writeHeader(parts: string[], def: DomainDef): void {
	parts.push(`Domain: ${def.name}\n`);
	parts.push(`Source: ${def.source}@${def.dsn}\n`);
}

// ─── Section: aliases ────────────────────────────────────────────────

/**
 * writeAliases emits German alias lines that boost retrieval for
 * queries phrased in German. The legacy renderer derived these from
 * `_list`/`_detail` suffixes on context names. The new domain DSL has
 * no such suffix — the domain is just `person` — so we emit a
 * uniform set that covers list, detail and edit phrasings.
 */
function writeAliases(parts: string[], def: DomainDef): void {
	// Skip the bare name — domainAliases includes it for the
	// resolver, but the chunk's "Alias:" line is supposed to add
	// retrieval-friendly variants on top of the name that already
	// appears in the header.
	const variants = domainAliases(def).filter((a) => a !== def.name);
	parts.push(`Alias: ${variants.join(", ")}\n`);
}

// ─── Section: fields ─────────────────────────────────────────────────

/** writeFields lists every field with type and filterable/readonly attrs. */
function writeFields(parts: string[], def: DomainDef): void {
	if (def.fields.length === 0) return;
	const descs = def.fields.map(describeField);
	parts.push(`Fields: ${descs.join(" | ")}\n`);

	// Compact "ALLOWED query fields" line — all field names, comma-
	// separated. Tells the LLM exactly which columns it may select.
	const names = def.fields.map((f) => f.name).join(", ");
	parts.push(`ALLOWED query fields (ONLY these, no others): ${names}\n`);

	// Canonical "fetch all" GraphQL query against the domain.
	const selection = def.fields.map((f) => f.name).join(" ");
	parts.push(`GraphQL query all: { ${def.name} { ${selection} } }\n`);
}

/** describeField renders one field as `name (type, attr1, attr2, ...)`. */
function describeField(f: DomainFieldDef): string {
	const attrs: string[] = [];
	if (f.readOnly) attrs.push("readonly");
	if (f.filterable) attrs.push("filterable");
	if (f.optionsRef) attrs.push(`options=${f.optionsRef}`);
	if (attrs.length === 0) {
		return `${f.name} (${f.type})`;
	}
	return `${f.name} (${f.type}, ${attrs.join(", ")})`;
}

// ─── Section: filter examples ────────────────────────────────────────

/**
 * writeFilterExamples emits filter examples for every filterable
 * field. Two passes per field: first the type-driven defaults (so
 * the LLM sees the syntax for every supported operator), then the
 * author overrides from `example` blocks (so the chunk feels
 * specific and accurate).
 */
function writeFilterExamples(parts: string[], def: DomainDef): void {
	const filterable = def.fields.filter((f) => f.filterable);
	if (filterable.length === 0) return;

	const names = filterable.map((f) => f.name).join(", ");
	parts.push(`Filterable fields: ${names}\n`);

	// The selection set used inside every example query. Using the
	// full field list is closer to what the LLM actually emits at
	// runtime than guessing a smaller "list_fields" subset.
	const selection = def.fields.map((f) => f.name).join(" ");

	for (const f of filterable) {
		// Pass 1: typed defaults.
		for (const op of operatorsForType(f.type)) {
			parts.push(renderFilterExampleLine(def.name, f, op, op.sampleValue, selection));
		}

		// Pass 2: author overrides.
		for (const ex of f.examples) {
			const op = findOperator(f.type, ex.op);
			if (!op) continue; // Unknown operator for this type — skip silently.
			const value = formatExampleValue(f.type, ex.value, ex.valueIsString);
			parts.push(renderFilterExampleOverride(def.name, f, op, ex, value, selection));
		}
	}

	// Multi-filter example: show how to combine two filters in one query.
	// Without this the LLM never sees the combined-args syntax and fails
	// when the user asks for two conditions at once.
	if (filterable.length >= 2) {
		parts.push(renderMultiFilterExample(def.name, filterable, selection));
	}
}

/**
 * renderMultiFilterExample picks the first two filterable fields and
 * combines their default operators into a single GraphQL query, e.g.:
 *   { person(age_gt: 0, city_eq: "value") { … } }
 *
 * This teaches the LLM that multiple filter args are comma-separated
 * inside the same parenthesis — not two separate queries.
 */
function renderMultiFilterExample(
	domainName: string,
	filterable: DomainFieldDef[],
	selection: string,
): string {
	const [a, b] = filterable;
	if (!a || !b) return "";
	const opA = operatorsForType(a.type)[0];
	const opB = operatorsForType(b.type)[0];
	if (!opA || !opB) return "";
	const argA = renderFilterArg(a.name, opA.suffix, opA.sampleValue);
	const argB = renderFilterArg(b.name, opB.suffix, opB.sampleValue);
	return (
		`Filter example (combined — ${a.name} and ${b.name}): ` +
		`{ ${domainName}(${argA}, ${argB}) { ${selection} } }\n` +
		`Filter note: multiple filters are comma-separated inside the same parenthesis. ` +
		`Any filterable field can be combined this way.\n`
	);
}

/**
 * renderFilterExampleLine formats one type-driven default: a single
 * filter argument with the catalog's sample value, paired with the
 * full selection set so the LLM always sees a complete query.
 */
function renderFilterExampleLine(
	domainName: string,
	f: DomainFieldDef,
	op: FilterOp,
	value: string,
	selection: string,
): string {
	const arg = renderFilterArg(f.name, op.suffix, value);
	return `Filter example (${f.name} ${op.label}): { ${domainName}(${arg}) { ${selection} } }\n`;
}

/**
 * renderFilterExampleOverride formats one author-supplied override.
 * Includes the description as a trailing comment so the embedding
 * model picks up the realistic phrasing.
 */
function renderFilterExampleOverride(
	domainName: string,
	f: DomainFieldDef,
	op: FilterOp,
	ex: ExampleDef,
	value: string,
	selection: string,
): string {
	const arg = renderFilterArg(f.name, op.suffix, value);
	const header = ex.description
		? `Filter example (${f.name} ${op.label}) — ${ex.description}`
		: `Filter example (${f.name} ${op.label})`;
	return `${header}: { ${domainName}(${arg}) { ${selection} } }\n`;
}

// ─── Section: relations ──────────────────────────────────────────────

/** writeRelations renders has_many / has_one / belongs_to lines. */
function writeRelations(parts: string[], def: DomainDef): void {
	if (def.relations.length === 0) return;
	const lines = def.relations.map(describeRelation);
	parts.push(`Relations: ${lines.join(" | ")}\n`);
}

/** describeRelation renders one relation in compact form. */
function describeRelation(r: RelationDef): string {
	return `${r.name} (${r.kind} ${r.target}, ${r.localField} -> ${r.foreignField})`;
}

// ─── Section: meta sources ───────────────────────────────────────────

/**
 * writeMetas emits everything an LLM needs to populate dropdown fields:
 *
 *   1. A "Dropdown sources" summary line listing every meta with its
 *      source table and label column (good for human readers and for
 *      retrieval against German queries like "alle Städte").
 *
 *   2. One ready-to-copy GraphQL meta query per meta, named verbatim
 *      after the meta — `meta_<name>` — so the LLM never has to
 *      construct the query name.
 *
 * Emitting (2) explicitly is the whole point of this block: it turns
 * an open-ended reasoning task into a template-filling task, which
 * smaller models handle reliably.
 */
function writeMetas(parts: string[], def: DomainDef): void {
	if (def.metas.length === 0) return;

	const summary = def.metas.map(describeMetaSource).join(" | ");
	parts.push(`Dropdown sources: ${summary}\n`);

	parts.push("Meta queries (copy verbatim, do not invent names):\n");
	for (const m of def.metas) {
		parts.push(`  - ${m.name}: { ${metaQueryName(m)} { value label } }\n`);
	}
}

/** describeMetaSource renders one meta as `name (from table, label=col)`. */
function describeMetaSource(m: MetaDef): string {
	return `${m.name} (from ${m.table}.${m.valueField}, label=${m.labelField})`;
}

/**
 * metaQueryName returns the canonical GraphQL meta-query name for a
 * Meta. The contract — `meta_<name>` — is fixed: the GraphQL backend
 * exposes meta queries under exactly this prefix, and the chunk
 * teaches the LLM that fact.
 */
function metaQueryName(m: MetaDef): string {
	return `meta_${m.name}`;
}

// ─── Section: dropdown field mapping ─────────────────────────────────

/**
 * writeDropdownFieldMapping emits two blocks that together teach the
 * LLM how to build a single GraphQL request that carries both the
 * record and every dropdown's options.
 *
 * "Dropdown fields" pairs each field that has an optionsRef with its
 * meta query name, so the mapping is unambiguous even when field and
 * meta don't share a name (e.g. field "role" → meta "roles" → query
 * "meta_roles").
 *
 * "Full example combined query" stitches the main domain query with
 * every meta query into a single brace-block. That's the exact shape
 * the LLM should emit when asked for a record that has dropdowns.
 */
function writeDropdownFieldMapping(parts: string[], def: DomainDef): void {
	const pairs = collectDropdownPairs(def);
	if (pairs.length === 0) return;

	parts.push(
		"Dropdown fields (every one below must be fetched together with its meta):\n",
	);
	for (const p of pairs) {
		parts.push(`  - ${p.field} -> ${p.query}\n`);
	}

	const recordSelection = def.fields.map((f) => f.name).join(" ");
	if (!recordSelection) return;

	// Deduplicate meta blocks: multiple fields can share a meta and
	// the combined query should only fetch each meta once.
	const seen = new Set<string>();
	const metaBlocks: string[] = [];
	for (const p of pairs) {
		if (seen.has(p.query)) continue;
		seen.add(p.query);
		metaBlocks.push(`${p.query} { value label }`);
	}

	parts.push(
		`Full example combined query: { ${def.name} { ${recordSelection} } ${metaBlocks.join(" ")} }\n`,
	);
}

/** Internal pair: the field name and the meta query it sources from. */
interface DropdownPair {
	field: string;
	query: string;
}

/**
 * collectDropdownPairs walks the domain's fields and returns one pair
 * per field whose optionsRef resolves to a declared meta. Fields with
 * a dangling optionsRef are dropped silently — a missing meta is a
 * seed problem, not a chunk problem.
 */
function collectDropdownPairs(def: DomainDef): DropdownPair[] {
	const queryByName = new Map<string, string>();
	for (const m of def.metas) {
		queryByName.set(m.name, metaQueryName(m));
	}

	const pairs: DropdownPair[] = [];
	for (const f of def.fields) {
		if (!f.optionsRef) continue;
		const q = queryByName.get(f.optionsRef);
		if (!q) continue;
		pairs.push({ field: f.name, query: q });
	}
	return pairs;
}

// ─── Section: permissions ────────────────────────────────────────────

/** writePermissions renders the role-based permission matrix. */
function writePermissions(parts: string[], def: DomainDef): void {
	if (def.permissions.length === 0) return;
	const lines = def.permissions.map(describePermission);
	parts.push(`Permissions: ${lines.join(" | ")}\n`);
}

/** describePermission renders one role's actions as `role=read,write,...`. */
function describePermission(p: PermissionDef): string {
	return `${p.role}=${p.actions.join(",")}`;
}

// ─── Section: AI hints ───────────────────────────────────────────────

/**
 * writeAiHints renders human-authored AI hints — behaviour rules,
 * format hints, caveats. Empty bodies are skipped silently so an
 * accidental empty block doesn't pollute the chunk.
 */
function writeAiHints(parts: string[], def: DomainDef): void {
	const usable = def.aiHints.filter((h) => collapseWhitespace(h.body) !== "");
	if (usable.length === 0) return;

	parts.push("AI hints:\n");
	for (const h of usable) {
		parts.push(`  - ${h.name}: ${collapseWhitespace(h.body)}\n`);
	}
}

/**
 * collapseWhitespace normalises runs of whitespace (newlines, tabs,
 * multiple spaces) into single spaces. Authors write multi-line AI
 * hints for readability; the embedding model wants flat text.
 */
function collapseWhitespace(s: string): string {
	return s.replace(/\s+/g, " ").trim();
}

// Re-export for tests / external consumers that want to inspect a
// single field's filter examples without rendering the whole chunk.
export type { FieldTypeDef, AiHintDef };

File: crates/oos-dsls/src/domain/llm_chunk.rs
//! DomainDef -> LLM-friendly text chunk.
//!
//! The chunk teaches the agent how to read a domain through the
//! `oos_query` tool, which calls oos.cmd.data.query: the field list, the
//! per-field filter operators, the {field, op, value} where shape, the
//! dropdown vocabularies, and the permission matrix. GraphQL is retired
//! — no query strings appear here.
//!
//! Deterministic: same DomainDef -> same string, byte for byte, so the
//! backfill can hash-and-skip unchanged chunks. The section order
//! (header, aliases, fields, filter examples, relations, metas, dropdown
//! mapping, permissions, ai hints) is part of the retrieval contract.

use std::collections::HashSet;

use super::aliases::domain_aliases;
use super::operators::{find_operator, format_example_value, operators_for_type, render_where_entry};
use super::types::{
    AiHintDef, DomainDef, DomainFieldDef, FieldType, MetaDef, PermissionAction, PermissionDef,
    RelationDef, RelationKind,
};

/// Renders the structured chunk embedded into pgvector for LLM retrieval.
pub fn render_llm_chunk(def: &DomainDef) -> String {
    let mut s = String::new();
    write_header(&mut s, def);
    write_aliases(&mut s, def);
    write_fields(&mut s, def);
    write_filter_examples(&mut s, def);
    write_relations(&mut s, def);
    write_metas(&mut s, def);
    write_dropdown_field_mapping(&mut s, def);
    write_permissions(&mut s, def);
    write_ai_hints(&mut s, def);
    s
}

// ─── header ───────────────────────────────────

fn write_header(s: &mut String, def: &DomainDef) {
    s.push_str(&format!("Domain: {}\n", def.name));
    s.push_str(&format!("Source: {}@{}\n", def.source, def.dsn));
}

// ─── aliases ───────────────────────────

/// Emits retrieval-friendly variants on top of the bare name, which
/// already appears in the header — so the name itself is filtered out.
fn write_aliases(s: &mut String, def: &DomainDef) {
    let variants: Vec<String> = domain_aliases(def)
        .into_iter()
        .filter(|a| a != &def.name)
        .collect();
    s.push_str(&format!("Alias: {}\n", variants.join(", ")));
}

// ─── fields ────────────────────────────

fn write_fields(s: &mut String, def: &DomainDef) {
    if def.fields.is_empty() {
        return;
    }
    let descs: Vec<String> = def.fields.iter().map(describe_field).collect();
    s.push_str(&format!("Fields: {}\n", descs.join(" | ")));

    let names: Vec<&str> = def.fields.iter().map(|f| f.name.as_str()).collect();
    s.push_str(&format!(
        "ALLOWED query fields (ONLY these, no others): {}\n",
        names.join(", ")
    ));

    // Read path: oos_query against this domain. Omitting `where` returns
    // every row; the result always carries the full field list above.
    s.push_str(&format!(
        "Read: call oos_query with context_name \"{}\" and an optional where array (omit it for all rows).\n",
        def.name
    ));
}

fn describe_field(f: &DomainFieldDef) -> String {
    let mut attrs: Vec<String> = Vec::new();
    if f.read_only {
        attrs.push("readonly".to_string());
    }
    if f.filterable {
        attrs.push("filterable".to_string());
    }
    if let Some(opt) = &f.options_ref {
        attrs.push(format!("options={opt}"));
    }
    let ty = field_type_str(f.field_type);
    if attrs.is_empty() {
        format!("{} ({})", f.name, ty)
    } else {
        format!("{} ({}, {})", f.name, ty, attrs.join(", "))
    }
}

fn field_type_str(t: FieldType) -> &'static str {
    match t {
        FieldType::Int => "int",
        FieldType::Float => "float",
        FieldType::String => "string",
        FieldType::Text => "text",
        FieldType::Bool => "bool",
        FieldType::Date => "date",
        FieldType::Datetime => "datetime",
    }
}

// ─── filter examples ──────────────────────────

fn write_filter_examples(s: &mut String, def: &DomainDef) {
    let filterable: Vec<&DomainFieldDef> = def.fields.iter().filter(|f| f.filterable).collect();
    if filterable.is_empty() {
        return;
    }
    let names: Vec<&str> = filterable.iter().map(|f| f.name.as_str()).collect();
    s.push_str(&format!("Filterable fields: {}\n", names.join(", ")));
    s.push_str(
        "Filter with a where array; each entry is { \"field\": <name>, \"op\": <op>, \"value\": <value> }. Multiple entries are AND-combined.\n",
    );

    // Allowed operator tokens per filterable field — the runtime rejects
    // any op not listed here for the field's type.
    let by_field: Vec<String> = filterable
        .iter()
        .map(|f| {
            let toks: Vec<&str> = operators_for_type(f.field_type).iter().map(|o| o.token).collect();
            format!("{} [{}]", f.name, toks.join(", "))
        })
        .collect();
    s.push_str(&format!("Allowed operators by field: {}\n", by_field.join(" | ")));

    // Concrete where examples: one typed default per field (its first
    // op), then any author-supplied overrides with their descriptions.
    for f in &filterable {
        if let Some(op) = operators_for_type(f.field_type).first() {
            let entry = render_where_entry(&f.name, op.token, op.sample_value);
            s.push_str(&format!("where example ({} {}): [{}]\n", f.name, op.label, entry));
        }
        for ex in &f.examples {
            let Some(op) = find_operator(f.field_type, ex.op) else {
                continue;
            };
            let value = format_example_value(f.field_type, &ex.value, ex.value_is_string);
            let entry = render_where_entry(&f.name, op.token, &value);
            let header = if ex.description.is_empty() {
                format!("where example ({} {})", f.name, op.label)
            } else {
                format!("where example ({} {}) — {}", f.name, op.label, ex.description)
            };
            s.push_str(&format!("{}: [{}]\n", header, entry));
        }
    }

    // Combined-filter example: two AND-ed entries in one where array.
    if filterable.len() >= 2 {
        if let Some(line) = render_multi_filter_example(&filterable) {
            s.push_str(&line);
        }
    }
}

fn render_multi_filter_example(filterable: &[&DomainFieldDef]) -> Option<String> {
    let a = filterable[0];
    let b = filterable[1];
    let op_a = operators_for_type(a.field_type).into_iter().next()?;
    let op_b = operators_for_type(b.field_type).into_iter().next()?;
    let entry_a = render_where_entry(&a.name, op_a.token, op_a.sample_value);
    let entry_b = render_where_entry(&b.name, op_b.token, op_b.sample_value);
    Some(format!(
        "where example (combined — {} and {}, AND): [{}, {}]\n",
        a.name, b.name, entry_a, entry_b
    ))
}

// ─── relations ─────────────────────────

fn write_relations(s: &mut String, def: &DomainDef) {
    if def.relations.is_empty() {
        return;
    }
    let lines: Vec<String> = def.relations.iter().map(describe_relation).collect();
    s.push_str(&format!("Relations: {}\n", lines.join(" | ")));
}

fn describe_relation(r: &RelationDef) -> String {
    format!(
        "{} ({} {}, {} -> {})",
        r.name,
        relation_kind_str(r.kind),
        r.target,
        r.local_field,
        r.foreign_field
    )
}

fn relation_kind_str(k: RelationKind) -> &'static str {
    match k {
        RelationKind::HasMany => "has_many",
        RelationKind::HasOne => "has_one",
        RelationKind::BelongsTo => "belongs_to",
    }
}

// ─── meta sources ───────────────────────

fn write_metas(s: &mut String, def: &DomainDef) {
    if def.metas.is_empty() {
        return;
    }
    let summary: Vec<String> = def.metas.iter().map(describe_meta_source).collect();
    s.push_str(&format!("Dropdown sources: {}\n", summary.join(" | ")));
}

fn describe_meta_source(m: &MetaDef) -> String {
    format!("{} (from {}.{}, label={})", m.name, m.table, m.value_field, m.label_field)
}

// ─── dropdown field mapping ────────────────────

fn write_dropdown_field_mapping(s: &mut String, def: &DomainDef) {
    let pairs = collect_dropdown_pairs(def);
    if pairs.is_empty() {
        return;
    }
    s.push_str(
        "Dropdown fields (closed vocabularies — values come only from the meta list, never free text):\n",
    );
    for p in &pairs {
        s.push_str(&format!("  - {} -> {}\n", p.field, p.meta));
    }
    s.push_str(
        "Option lists are returned by oos_query when called with withOptions: true, in `options` keyed by meta name.\n",
    );
}

struct DropdownPair {
    field: String,
    meta: String,
}

fn collect_dropdown_pairs(def: &DomainDef) -> Vec<DropdownPair> {
    let meta_names: HashSet<&str> = def.metas.iter().map(|m| m.name.as_str()).collect();
    let mut pairs = Vec::new();
    for f in &def.fields {
        if let Some(opt) = &f.options_ref {
            if meta_names.contains(opt.as_str()) {
                pairs.push(DropdownPair { field: f.name.clone(), meta: opt.clone() });
            }
        }
    }
    pairs
}

// ─── permissions ───────────────────────

fn write_permissions(s: &mut String, def: &DomainDef) {
    if def.permissions.is_empty() {
        return;
    }
    let lines: Vec<String> = def.permissions.iter().map(describe_permission).collect();
    s.push_str(&format!("Permissions: {}\n", lines.join(" | ")));
}

fn describe_permission(p: &PermissionDef) -> String {
    let actions: Vec<&str> = p.actions.iter().map(|a| permission_action_str(*a)).collect();
    format!("{}={}", p.role, actions.join(","))
}

fn permission_action_str(a: PermissionAction) -> &'static str {
    match a {
        PermissionAction::Read => "read",
        PermissionAction::Write => "write",
        PermissionAction::Delete => "delete",
    }
}

// ─── AI hints ────────────────────────

fn write_ai_hints(s: &mut String, def: &DomainDef) {
    let usable: Vec<&AiHintDef> = def
        .ai_hints
        .iter()
        .filter(|h| !collapse_whitespace(&h.body).is_empty())
        .collect();
    if usable.is_empty() {
        return;
    }
    s.push_str("AI hints:\n");
    for h in &usable {
        s.push_str(&format!("  - {}: {}\n", h.name, collapse_whitespace(&h.body)));
    }
}

/// Normalises runs of whitespace to single spaces and trims — authors
/// write multi-line hints for readability; the embedder wants flat text.
fn collapse_whitespace(s: &str) -> String {
    s.split_whitespace().collect::<Vec<_>>().join(" ")
}

#[cfg(test)]
mod tests {
    use super::super::parse_domain;
    use super::render_llm_chunk;

    const NOTE_DOMAIN: &str = r#"
domain note from note@demo {
  permission admin   read, write, delete
  permission manager read, write
  permission user    read

  field id         : int      readonly
  field person_id  : int      readonly filterable {
    example eq 42 "Alle Notizen der Person mit id 42"
  }
  field title      : string   filterable {
    example like "Meeting" "Titel enthält 'Meeting'"
  }
  field body       : text
  field created_at : datetime readonly

  ai "scope"
     "Notizen gehören immer zu einer Person."

  aliases [ "notizen", "anmerkungen" ]
}
"#;

    #[test]
    fn renders_note_chunk_sections() {
        let def = parse_domain(NOTE_DOMAIN).expect("note domain parses");
        let chunk = render_llm_chunk(&def);

        // Header + identity.
        assert!(chunk.starts_with("Domain: note\nSource: note@demo\n"));
        // Author alias expanded into the German phrasing family.
        assert!(chunk.contains("Alias: "));
        assert!(chunk.contains("notizen"));
        // Field listing + the ALLOWED line the LLM is told to obey.
        assert!(chunk.contains("ALLOWED query fields (ONLY these, no others): id, person_id, title, body, created_at"));
        // Read path is oos_query / data.query, not GraphQL.
        assert!(chunk.contains("Read: call oos_query with context_name \"note\""));
        assert!(!chunk.contains("GraphQL"));
        // Filterable fields + per-field op tokens (data.query vocabulary).
        assert!(chunk.contains("Filterable fields: person_id, title"));
        assert!(chunk.contains("Allowed operators by field: person_id [eq, ne, gt, gte, lt, lte] | title [contains, eq, ne]"));
        // Typed-default where example + the author override (bare int value).
        assert!(chunk.contains("where example (person_id equals): [{ \"field\": \"person_id\", \"op\": \"eq\", \"value\": 0 }]"));
        assert!(chunk.contains("Alle Notizen der Person mit id 42"));
        assert!(chunk.contains("\"value\": 42"));
        // String author override quotes the value.
        assert!(chunk.contains("where example (title contains) — Titel enthält 'Meeting'"));
        // Combined where example (two filterable fields present).
        assert!(chunk.contains("where example (combined — person_id and title, AND):"));
        // Permissions matrix.
        assert!(chunk.contains("Permissions: admin=read,write,delete | manager=read,write | user=read"));
        // AI hint, whitespace-collapsed.
        assert!(chunk.contains("AI hints:"));
        assert!(chunk.contains("  - scope: Notizen gehören immer zu einer Person."));
    }
}
